### PR TITLE
#942 process colorsWithoutAlpha

### PR DESCRIPTION
**Pull request by andreicristianpetcu**
image

### CODE REVIEW

**Comment by andreicristianpetcu** [comment]
Example of a theme using these values (the original classic theme I wanted to port) http://localhost:8080/?theme=XQAAAAJ4BAAAAAAAAABBKYhm849SCicxcUhA3DJozHnOMuotJJIF9jJb-FWmo1SjM2MWLJWW2J5lCZzVP2MQDfAs41MN24hV_0CG0VmlorOZhd96gWR3F52l7wx3zlkkdyKtrV34f6wGULgIRdBYEhR1tpgWvD_P4ZiLWaBEAmloHI_5cxZ9x_cZM_OK88Nl3ZIJe2eS5nyeP8CVtq6NY7LYt67RAbjyabSi8PF-jh1xB-IZixQYUZSZxctWZBDsSopT0DfLlnYr4w6gyAVpf9LjPKmpzX5rhrRh5uLFeKeBveGuNoSVr_IZMKN_fC2jQv7MfdzRvP3RsVULqjnkwGsL2-yCinV5zVwgUSUmTSRh5DTaJWilqNxV1pkqmS-E-dmMAivIUwQLh6V6vdg87cg3Dyu3-gB-sN9zX-Yeb9L8i8PPVDV1jdc4263Vi9iMKS8NE-LhtuFQz_4Lo1Z9Y3ObhijmoTYggwfBNTsMHKMXReSAeh7qpRYNzwg0Nq4ygDrWZRVtOcc6Rxw-o9DlEMMdoAWTpkZghrdIotT1W2Yh_aQXJHDiQy86aHrjUT3mZ0ql56Hc6JUFpfB5-4nGcw

**Comment by andreicristianpetcu** [comment]
Damn.... there is a validator that forces a color to either have or not have alpha for all the themes. What's the point of having `a: 1`?
https://app.circleci.com/pipelines/github/mozilla/FirefoxColor/1015/workflows/1e2bfbc9-ac1a-49bb-ad2c-4bec381a504c/jobs/2707
image

**Comment by Rob--W** [comment]
Interesting, I wasn't aware of `docs/theme-schema.json` before. Could you update it to add all new colors that you added in the advanced colors PR?

And rather than "fixing" the theme defaults by stripping the a:1, change the definition to make "a" optional.

**Comment by andreicristianpetcu** [comment]
done @Rob--W

File: bin/convert-themes-to-rgb.js [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 
 const makeTinycolor = colorIn => {
   let { a, s } = colorIn;
-  return tinycolor({
+  let newColor = tinycolor({
     ...colorIn,
-    a: Math.floor(a) / 100.0,
     s: Math.floor(s) / 100.0
   });
+  if (a) {
```
Comment (Rob--W) [inline bin/convert-themes-to-rgb.js:13]: `a != undefined` to make sure that if `a: 0` is used, that it continues to be like that.

And please add a unit test for this case.

File: src/lib/themes.js [hunk 2]
```diff
@@ -125,8 +125,9 @@ export const normalizeThemeBackground = background =>
 
 // Utility to ensure normal & consistent colors
 export const normalizeThemeColor = (name, data, defaultColor) => {
-  const color = makeTinycolor(data || defaultColor).toRgb();
-  if (!colorsWithAlpha.includes(name)) {
+  const inColor = data || defaultColor;
+  const color = makeTinycolor(inColor).toRgb();
+  if (colorsWithoutAlpha.includes(name) || (inColor && !("a" in inColor))) {
```
Comment (Rob--W) [inline src/lib/themes.js:130]: If alpha has the default value, it can also be deleted IMO.

Could you add unit tests to cover this new logic?

Comment (andreicristianpetcu) [reply src/lib/themes.js:130]: I'm a bit confused. I initially proposed this.
image
https://github.com/mozilla/FirefoxColor/pull/944#issuecomment-663971106
Then I implemented this change. https://github.com/mozilla/FirefoxColor/commit/b911023c3d60918866306d3b15b19a80ae7d6b88#diff-28b54e29959f783d6744df0f4e0c3c49R28
image
Then you told me to revert. https://github.com/mozilla/FirefoxColor/pull/944#issuecomment-664003245
image

I'm fine with either way but if we delete the opaque alpha, then the "[should not change preset themes](https://github.com/mozilla/FirefoxColor/blob/410e1eda9f6cc9a4837edefef136398712619c59/src/lib/themes-test.js#L27-L37)" will fail if we don't update the preset themes like I did before.
image

Am I missing some solution where we can delete the `a: 1` but we avoid updating the default themes?

File: bin/convert-themes-to-rgb.js (hunk 1 above)
Comment (andreicristianpetcu) [reply bin/convert-themes-to-rgb.js:13]: Where is this called from? Should I create a new test for it?

File: src/lib/themes.js (hunk 2 above)
Comment (Rob--W) [reply src/lib/themes.js:130]: I thought that you deleted `"a": 1` in order to get it to pass the schema validation.

I'd be fine with removing `"a": 1` from the default theme templates if it is still backwards-compatible.
Could you verify that the website would still be compatible with the old add-on (i.e. the current 2.1.6) if you do remove `"a": 1`?

File: bin/convert-themes-to-rgb.js (hunk 1 above)
Comment (Rob--W) [reply bin/convert-themes-to-rgb.js:13]: Oh this is not called from anywhere, it's a tool. Never mind the unit test for this file.

You should be able to undo the change though, as the `a` property is deleted below depending on conditions.

File: src/lib/themes.js (hunk 2 above)
Comment (andreicristianpetcu) [reply src/lib/themes.js:130]: do we get any benefit if we delete the a: 1?

Comment (Rob--W) [reply src/lib/themes.js:130]: I suppose that the permalink to the URL would be smaller (i.e. the part after `?theme=...`).

**Comment by andreicristianpetcu** [comment]
@Rob--W here is a URL of the first theme which by default has alpha set to 1 in the toolbar. This URL does not have any alpha. It works fine. https://color.firefox.com/?theme=XQAAAAIwAQAAAAAAAABBqYhm849SCia2CaaEGccwS-xMDPr6QjyuAiEifwJmpSs38LXYwAM0txHvymCyYe8A373Y-5dI6pwZVVCPpCb6CSC650B4-wuqUzhU9zT1DXDh8Cl7n78o7Ucx1t-v-XJHZv9mutHz9tBgdjJSgPDugfulzDxfznqKATnRv7bV9YYnrnpWmkZxS6IwaDDhgLClbtw0iNx1VqSMyOyILuNgYmNImKhHxo2zNo7e43RFZJPqljK0zz_7I6HP6zqBQjyXvPat4nXmJl2M3if_jyCAAA

File: src/lib/themes.js [hunk 3]
```diff
@@ -89,15 +89,16 @@ export const themesEqual = (themeA, themeB) => {
...
-      return false;
-    }
-    if (
-      alphaInA &&
-      alphaInB &&
-      Math.abs(colorA.a - colorB.a) > alphaEqualityTolerance
-    ) {
-      return false;
-    }
+      if ((alphaInA && colorA.a !== 1) || (alphaInB && colorB.a !== 1)) {
```
Comment (Rob--W) [inline src/lib/themes.js:92]: The logic of these conditions spread over multiple lines look complicated.

Since absence of `a` is now treated like `a: 1`, use the nullish coalescing operator to get the desired (numeric) value for `alphaA` and `alphaB`. Then you'll only need one if-condition like `if (Math.abs(alphaA - alphaB) > alphaEqualityTolerance) {`

Note that I'm suggesting the nullish coalescing operator instead of OR, to make sure that `0` stays `0` (i.e. transparent) and doesn't become `1` (opaque). Considering these edge cases, it would also be useful to check if there are existing tests for alpha being undefined, 0, 1, in between (e.g. 0.5), and if they don't exist yet, add them.

**Comment by andreicristianpetcu** [comment]
Done @Rob--W :ballot_box_with_check:

File: src/lib/themes-test.js [hunk 4]
```diff
@@ -300,6 +309,33 @@ describe("lib/themes", () => {
...
+    it("should accept if both alpha are set to opaque", () => {
+      expect(
+        subject(
+          { colors: { toolbar: { r: 1, g: 2, b: 3, a: 1 } }, },
+          { colors: { toolbar: { r: 1, g: 2, b: 3, a: 1 } } }
+        )
+      ).to.be.true;
+    });
+
+    it("should accept if both alpha are set to transparent", () => {
```
Comment (Rob--W) [inline src/lib/themes-test.js:330]: Those are "good weather" tests. If you change the implementation below from `?? 1` to `|| 1`, will you see any test failure? From what I have seen in the tests, the answer appears to be no.

To get decent coverage, you should at the very leasy also be comparing `a: 0` with `a: 1` (expect false) and `a: 0` with `(no a)` (expect false).

And since the source code contains the (hardcoded) magic number `1`, it might be a good idea to compare these boundary values 0/1 with a different value in between (e.g. 0.5): compare `a: 0` with `a: 0.5`, and `a: 1` with `a: 0.5`.

**Comment by andreicristianpetcu** [comment]
done @Rob--W :rocket:

**Comment by andreicristianpetcu** [comment]
Did you have time to submit the extension to AMO, @Rob--W ?

**Comment by Rob--W** [comment]
Not yet. I started with some testing last week, found a few issues and fixed them (I just pushed them to master). I haven't had the chance to complete the tests before release.

**Comment by andreicristianpetcu** [comment]
@Rob--W Were you affected by the layoffs? :(

**Comment by Rob--W** [comment]
@andreicristianpetcu Thanks for your concern, I'm not.

On this issue, I was trying to update all dependencies to the latest version (`npm audit fix`), but that broke and I haven't looked back yet.

**Comment by andreicristianpetcu** [comment]
Are you still stuck with the dependencies? Need help? @Rob--W

**Comment by Rob--W** [comment]
:_: Turns out that `npm audit fix` failed because I inadvertently put `package-lock=false` in `.npmrc` somewhere. After removing that `npm audit fix` ran to completion. I've opened a PR at #948 and if nothing fails I'll merge that. I'll need to check that the generated build output (before and after #948) is equivalent and not causing regressions.

**Comment by andreicristianpetcu** [comment]
Any update on the next version in amo @Rob--W ?

**Comment by andreicristianpetcu** [comment]
This got merged 2 months ago. Any ne version in sight? @Rob--W ?

**Comment by Rob--W** [comment]
The dependency update was done in #955, and I have bumped the version to 2.1.7.
The add-on is currently pending review on AMO. I'll notify you when the update has been rolled out.

**Comment by Rob--W** [comment]
The add-on has been approved and I've pushed the latest version of the website to production.

@andreicristianpetcu Thanks again for your contributions and patience.
